proxy/request_test.go

```diff
@@ -51,3 +56,97 @@ func TestPing(t *testing.T) {
 		}
 	}
 }
+
+func TestSendErrors(t *testing.T) {
+	var respObj jsonResponse
+	client := MatrixClient{}
+
+	var req string
+	var resp []byte
+
+	// no room_id
+	req = `{"id": "1234", "method": "send"}`
```

Comment:
A fairly common idiom for this kind of test, where you're doing very similar things over several examples of slightly-bad code is:

type testcase struct {
  req string
  respSubstring string
}
for _, tc := range []testcase{{
  `{"id": "1234", "method": "send"}`,
  `missing room_id`
}, {
  `{"id": "1234", "method": "send", "params": {"room_id": "xyz"}}`,
  `missing type`
}} {
  resp := handleRequest([]byte(tc.req), &client)
  if err := json.Unmarshal(resp, &respObj); err != nil {
    t.Error(tc.respSubstring, "JSON error", err)
    continue
  }
  if !strings.Contains(respObj.Error.Error, tc.respSubstring) {
    t.Error(tc.respSubstring, "Wrong error", respObj.Error.Error
  }
}